plugins/IPBFormatter/class.ipbformatter.plugin.php

```diff
@@ -141,22 +141,71 @@ public function nbbc() {
         return $this->_NBBC;
     }
 
+    /**
+     * Build an array of attachment records from the Media table, using the MediaID as each record's index.
+     *
+     * @return array|null
+     */
     public function media() {
         if ($this->_Media === null) {
-            try {
-                $i = Gdn::pluginManager()->getPluginInstance('FileUploadPlugin', Gdn_PluginManager::ACCESS_CLASSNAME);
-                $m = $i->mediaCache();
-            } catch (Exception $ex) {
-                $m = array();
+            // Set _Media to a non-null value, so we only do this once.
+            $this->_Media = array();
+
+            // Fire up a basic model, configured for the Media table
+            $mediaModel = new Gdn_Model('Media');
+
+            // Grab a reference to the instance of the current controller
+            $controller = Gdn::Controller();
+
+            // Grab the current discussion ID from the current controller
+            $discussionID = $controller->data('Discussion.DiscussionID');
+
+            // Grab comment data set from the current controller and verify it's populated
+            $comments = $controller->Data('Comments');
+            $commentIDs = array();
+            if ($comments instanceof Gdn_DataSet && $comments->NumRows()) {
+                // Build a collection of comment IDs
+                while ($currentComment = $comments->NextRow()) {
+                    $commentIDs[] = $currentComment->CommentID;
+                }
+                unset($currentComment);
             }
 
-            $media = array();
-            foreach ($m as $key => $data) {
-                foreach ($data as $row) {
-                    $media[$row->MediaID] = $row;
+            /**
+             * Select all media records with a ForeignID matching the discussion ID and a
+             * ForeignTable value of "discussion"
+             */
+            $mediaStatement = $mediaModel->SQL
+                ->Select('m.*')
+                ->From('Media m')
+                ->BeginWhereGroup()
+                ->Where('m.ForeignID', $discussionID)
+                ->Where('m.ForeignTable', 'discussion')
+                ->EndWhereGroup();
+
+            // If we have any comment IDs, find attachments related to them too
+            if (!empty($commentIDs)) {
```

Comment:
I benchmarked the following queries and averaged out their times after ten executions.  There were 30 matching media records for discussions and 138 matching records for comments.  These tests indicate an overall slower execution when separated into separate queries.

Single query using OR: *1.5ms*
```sql
select sql_no_cache m.*
from GDN_Media m
where (m.ForeignTable = 'discussion' and m.ForeignID = '119568')
	or (m.ForeignTable = 'comment' and m.ForeignID in ('54039', '664003', '854255'));
```

Dedicated discussion query: *0.9ms*
```sql
select sql_no_cache m.*
from GDN_Media m
where (m.ForeignTable = 'discussion' and m.ForeignID = '119568');
```

Dedicated comment query: *1.3ms*
```sql
select sql_no_cache m.*
from GDN_Media m 
where (m.ForeignTable = 'comment' and m.ForeignID in ('54039', '664003', '854255'))
```